### PR TITLE
blob: Add button to download file

### PR DESCRIPTION
**Pull request by keegancsmith**
This uses our existing "raw" endpoint to download a file. It is displayed as
part of the BlobPage.

image

Fixes https://github.com/sourcegraph/sourcegraph/issues/5478

### CODE REVIEW

**Comment by keegancsmith** [comment]
Help please: This doesn't fully work yet. It updates the location using what I assume some SPA router, rather than actually causing the browser to visit the URL. How do I fix that? I assume either don't use a `<*Link` or add something special to a router somewhere.

Also this has no tests. Advice if necessary?

**Comment by eseliger** [comment]
A native <a with download attribute should work :) I can pick it up if you like though

**Comment by keegancsmith** [comment]
> A native <a with download attribute should work :) I can pick it up if you like though

yes please!

**Comment by eseliger** [comment]
fixed, it now triggers a download directly

File: web/src/repo/blob/GoToRawAction.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+/**
+ * A repository header action that replaces the blob in the URL with the raw URL.
+ */
+export class GoToRawAction extends React.PureComponent<Props> {
+    public render(): JSX.Element {
+        const to = `/${encodeRepoRev(this.props.repoName, this.props.rev)}/-/raw/${this.props.filePath}`
+        return (
+            <a href={to} className="nav-link" data-tooltip="Raw (download file)" download={true}>
```
Comment (felixfbecker) [inline web/src/repo/blob/GoToRawAction.tsx]: Proposal: make the tooltip only "Download file"

File: web/src/repo/blob/GoToRawAction.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    rev?: string
+    filePath: string
+}
+
+/**
+ * A repository header action that replaces the blob in the URL with the raw URL.
+ */
+export class GoToRawAction extends React.PureComponent<Props> {
+    public render(): JSX.Element {
+        const to = `/${encodeRepoRev(this.props.repoName, this.props.rev)}/-/raw/${this.props.filePath}`
```
Comment (felixfbecker) [inline web/src/repo/blob/GoToRawAction.tsx]: This won't work if the file path has spaces or special characters in it, needs to be URL encoded

**Comment by felixfbecker** [comment]
I've wanted this feature for so long!

**Pull request merged**